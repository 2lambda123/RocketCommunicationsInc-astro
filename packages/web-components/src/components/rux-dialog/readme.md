# rux-modal

A Rux Modal interrupts the app experience to prompt a user to confirm an action or acknowledge a piece of information. It displays information along with a set of buttons allowing users to “Accept or Cancel” the actions presented within the Rux Modal.

## Guidelines

-   [Astro UXDS: Rux Modal](https://astrouxds.com/ui-components/dialog-box)

## Web Components Usage

### 1. Installation

#### Download the [Astro UXDS Component Library](https://github.com/RocketCommunicationsInc/astro-components) source to your project.

Via CLI:

```sh
git clone https://github.com/RocketCommunicationsInc/astro-components.git
```

Or, [download the Astro UXDS Components as a .zip](https://github.com/RocketCommunicationsInc/astro-components/archive/master.zip)

### 2. Import the Astro Rux Modal Web Component

This example assumes you're using the NPM package in `node_modules`. Otherwise, import the component using the path to the Astro Components directory in your project.

```javascript
import { RuxModal } from '@astrouxds/rux-modal/rux-modal.js'
```

### 3. Render the Astro Rux Modal Web Component

At a minimum, Astro UXDS Rux Modal requires a message. In this instance, a single button labeled "Cancel" will be automatically rendered.

```xml
<rux-modal message="This is a message"></rux-modal>
```

Pass properties as attributes of the Astro Rux Modal custom element:

```xml
<rux-modal
  title="Modal title"
  message="Modal message"
  confirmText="Ok"
  denyText="Cancel"
  customEvent="listen-for-me">
</rux-modal>
```

<!-- Auto Generated Below -->


## Properties

| Property       | Attribute       | Description                  | Type                  | Default     |
| -------------- | --------------- | ---------------------------- | --------------------- | ----------- |
| `confirmText`  | `confirm-text`  | Text for confirmation button | `string`              | `'Confirm'` |
| `denyText`     | `deny-text`     | Text for close button        | `string`              | `'Cancel'`  |
| `modalMessage` | `modal-message` | Dialog body message          | `string \| undefined` | `undefined` |
| `modalTitle`   | `modal-title`   | Dialog header title          | `string \| undefined` | `undefined` |
| `open`         | `open`          | Shows and hides dialog       | `boolean`             | `false`     |


## Events

| Event             | Description                            | Type                           |
| ----------------- | -------------------------------------- | ------------------------------ |
| `ruxdialogclosed` | Event that is fired when dialog closes | `CustomEvent<boolean \| null>` |
| `ruxdialogopened` | Event that is fired when dialog opens  | `CustomEvent<void>`            |


## Slots

| Slot          | Description                     |
| ------------- | ------------------------------- |
| `"(default)"` | the dialog's message or content |
| `"footer"`    | the footer of the dialog        |
| `"header"`    | the header of the dialog        |


## Shadow Parts

| Part               | Description                 |
| ------------------ | --------------------------- |
| `"confirm-button"` | the dialog's confirm button |
| `"container"`      | the dialog container        |
| `"deny-button"`    | the dialog's deny button    |
| `"dialog"`         | the native dialog element   |
| `"footer"`         | the footer of the dialog    |
| `"header"`         | the header of the dialog    |
| `"message"`        | the message of the dialog   |


## CSS Custom Properties

| Name                       | Description                         |
| -------------------------- | ----------------------------------- |
| `--modal-background-color` | [DEPRECATED] Modal background color |
| `--modal-border-color`     | [DEPRECATED] Modal border color     |
| `--modal-title-color`      | [DEPRECATED] Modal title color      |


## Dependencies

### Depends on

- [rux-button-group](../rux-button-group)
- [rux-button](../rux-button)

### Graph
```mermaid
graph TD;
  rux-dialog --> rux-button-group
  rux-dialog --> rux-button
  rux-button --> rux-icon
  style rux-dialog fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
