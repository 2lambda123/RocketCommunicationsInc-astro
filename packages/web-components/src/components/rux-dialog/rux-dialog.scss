@use '../../global/mixins';

:host {
    display: contents;
}

*,
*::before,
*::after {
    box-sizing: inherit;
}

//  this styling is only used in storybook up for presentational purposes
//  default flex position takes up the entire body element en spills overlay
//  on top of the docs page

:host([hidden]) {
    display: none;
}

rux-button-group {
    margin-top: auto;
    margin-left: auto;
    align-self: flex-end;
}

.rux-dialog {
    &__wrapper {
        position: fixed;
        top: 0;
        left: 0;
        height: 100vh;
        width: 100vw;
        display: flex;
        justify-content: center;
        align-items: center;
        z-index: 1100;
        background-color: rgba(0, 0, 0, 0.5);
        @include mixins.animation(fadeIn, 0.5s);
    }
    &__dialog {
        position: relative;
        display: flex;
        flex-direction: column;
        justify-content: space-between;
        background-color: var(--color-background-surface-default);
        width: 28rem;
        border: 0;
        margin: auto;
        padding: 0;
        user-select: none;
        box-shadow: var(--shadow-outer-dialog);
        max-height: 100%;
    }

    &__header {
        display: flex;
        flex-grow: 0;
        flex-shrink: 0;
        align-items: center;
        width: 100%;
        padding: 16px 20px;
        background-color: var(--color-background-surface-header);
        color: var(--color-text-primary);
        user-select: none;
        font-family: var(--font-heading-2-font-family);
        font-size: var(--font-heading-2-font-size);
        font-weight: var(--font-heading-2-font-weight);
        letter-spacing: var(--font-heading-2-letter-spacing);
        ::slotted(h6),
        ::slotted(h5),
        ::slotted(h4),
        ::slotted(h3),
        ::slotted(h2),
        ::slotted(h1) {
            margin-block-start: 0;
            margin-block-end: 0;
        }
    }

    &__content {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        padding: 1rem 1rem 0 1rem;
        color: var(--color-text-primary);
        overflow: auto;
        white-space: nowrap;
        &::-webkit-scrollbar {
            width: 16px;
            height: 16px;
            background-color: transparent;
        }

        &::-webkit-scrollbar-thumb {
            background-color: var(
                --color-border-interactive-muted,
                rgb(43, 101, 155)
            );
            border-radius: 8px;
            border: 3px solid transparent;
            background-clip: padding-box;
        }

        /* visually "centers" because the dark edge of the shadow gives the illusion this is offset */
        &::-webkit-scrollbar-thumb:vertical {
            border-left-width: 4px;
        }

        &::-webkit-scrollbar-thumb:horizontal {
            border-top-width: 4px;
        }

        &::-webkit-scrollbar-thumb:active,
        &::-webkit-scrollbar-thumb:hover {
            background-color: var(
                --color-background-interactive-default,
                rgb(58, 129, 191)
            );
        }

        &::-webkit-scrollbar-track,
        &::-webkit-scrollbar-corner {
            background-color: var(--color-background-surface-default);
        }

        &::-webkit-scrollbar-track:vertical {
            box-shadow: inset 3px 3px 3px 0px rgba(0, 0, 0, 0.5);
        }

        &::-webkit-scrollbar-track:horizontal {
            box-shadow: inset 1px 3px 3px 0px rgba(0, 0, 0, 0.5);
        }
    }
    &__footer {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        padding: 0 1rem 1rem 1rem;
        color: var(--color-text-primary);
    }

    &__message {
        margin: 0.5rem 1.875rem 2.5rem 1.875rem;
    }
    .rux-button {
        box-shadow: none !important;
    }
    &__header.hidden {
        display: none;
    }
    &__message.hidden {
        display: none;
    }
}

rux-icon {
    margin-right: 0.75rem;
}

@include mixins.keyframes(fadeIn) {
    0% {
        opacity: 0;
    }
    100% {
        opacity: 1;
    }
}
