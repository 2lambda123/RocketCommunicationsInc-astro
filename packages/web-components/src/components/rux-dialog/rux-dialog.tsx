import {
    Component,
    h,
    Prop,
    Event,
    EventEmitter,
    Element,
    Listen,
    Watch,
    Host,
    State,
    Fragment,
} from '@stencil/core'
import { hasSlot } from '../../utils/utils'

/**
 * @part container - the dialog container
 * @part dialog - the native dialog element
 * @part header - the header of the dialog
 * @part message - the message of the dialog
 * @part confirm-button - the dialog's confirm button
 * @part deny-button - the dialog's deny button
 * @part footer - the footer of the dialog
 *
 * @slot header - the header of the dialog
 * @slot (default) - the dialog's message or content
 * @slot footer - the footer of the dialog
 */
@Component({
    tag: 'rux-dialog',
    styleUrl: 'rux-dialog.scss',
    shadow: { delegatesFocus: true },
})
export class RuxDialog {
    /**
     * Shows and hides dialog
     */
    @Prop({ reflect: true, mutable: true }) open: boolean = false
    /**
     * Dialog body message
     */
    @Prop() message?: string
    /**
     * Dialog header title
     */
    @Prop() header?: string
    /**
     * Text for confirmation button
     */
    @Prop() confirmText: string = 'Confirm'
    /**
     * Text for close button
     */
    @Prop() denyText: string = 'Cancel'

    /**
     * Allows dialog to close when clicking off it
     */
    @Prop({ attribute: 'click-to-close' }) clickToClose: boolean = false

    /**
     * Event that is fired when dialog opens
     */
    @Event({
        eventName: 'ruxdialogopened',
        composed: true,
        bubbles: true,
    })
    ruxDialogOpened!: EventEmitter<void>
    /**
     * Event that is fired when dialog closes. If dialog is closed by clicking on the default confirm or deny buttons (when no footer slot is provided), then true or false will be emitted respectively on the event.detail.
     */
    @Event({
        eventName: 'ruxdialogclosed',
        composed: true,
        bubbles: true,
    })
    ruxDialogClosed!: EventEmitter<boolean | null>

    @Element() element!: HTMLRuxDialogElement

    @State() hasFooter = hasSlot(this.element, 'footer')
    @State() hasHeader = hasSlot(this.element, 'header')
    @State() hasMessage = hasSlot(this.element)
    private _userInput: boolean | null = null

    // confirm dialog if Enter key is pressed
    @Listen('keydown', { target: 'window' })
    handleKeyDown(ev: KeyboardEvent) {
        // prevent this from running if the slots version is being used
        if (this.open && !this.hasFooter) {
            const btns: NodeListOf<HTMLRuxButtonElement> = this.element.shadowRoot!.querySelectorAll(
                'rux-button'
            )
            if (ev.key === 'Enter') {
                //If enter is hit but the cancel/deny button is focused, we want to click that instead.
                let activeEl: any = this.element.shadowRoot?.activeElement
                if (activeEl && activeEl === btns[0]) {
                    this._userInput = false
                    btns[0].click()
                } else {
                    const button = this._getDefaultButton()
                    if (button) {
                        this._userInput = true
                        button.click()
                    }
                }
            }
            if (ev.key === 'Escape') {
                this._userInput = false
                btns[0].click()
            }
        }
    }

    // close dialog if click happens outside of dialog
    @Listen('click', { target: 'window' })
    handleClick(ev: MouseEvent) {
        if (this.clickToClose) {
            const wrapper = this._getWrapper()
            if (ev.composedPath()[0] === wrapper) {
                this.open = false
            }
        }
    }

    @Watch('open')
    handleOpen(isOpen: boolean) {
        if (isOpen && !this.hasFooter) {
            setTimeout(() => {
                const button = this._getDefaultButton()
                if (button) {
                    button.focus()
                }
            }, 0)
        }
        this.open
            ? this.ruxDialogOpened.emit()
            : this.ruxDialogClosed.emit(this._userInput)
    }

    private _handleDialogChoice(e: MouseEvent) {
        // convert string value to boolean
        const target = e.currentTarget as HTMLElement
        const choice = target.dataset.value === 'true'
        this._userInput = choice
        this.open = false
    }

    private _getDefaultButton(): HTMLElement | null {
        const buttonSet = this.element?.shadowRoot?.querySelectorAll(
            'rux-button:not([hidden])'
        ) as NodeListOf<HTMLElement>

        if (buttonSet.length > 0) {
            const defaultButton = buttonSet[buttonSet.length - 1]
            const shadow = defaultButton.shadowRoot?.querySelector('button')

            if (shadow) return shadow
        }

        return null
    }

    private _getWrapper(): HTMLElement | null {
        const wrapper = this.element?.shadowRoot?.querySelector(
            '.rux-dialog__wrapper'
        ) as HTMLElement

        if (wrapper) {
            return wrapper
        }
        return null
    }

    connectedCallback() {
        this._handleDialogChoice = this._handleDialogChoice.bind(this)
        this._handleSlotChange = this._handleSlotChange.bind(this)
    }

    componentDidRender() {
        const button = this._getDefaultButton()
        button && button.focus()
    }

    private _handleSlotChange() {
        this.hasHeader = hasSlot(this.element, 'header')
        this.hasMessage = hasSlot(this.element)
        this.hasFooter = hasSlot(this.element, 'footer')
    }

    render() {
        const {
            open,
            message,
            header,
            confirmText,
            denyText,
            _handleDialogChoice,
        } = this

        return (
            open && (
                <Host>
                    <div part="container" class="rux-dialog__wrapper">
                        <dialog
                            class="rux-dialog__dialog"
                            role="dialog"
                            part="dialog"
                        >
                            <header
                                class={{
                                    hidden:
                                        !this.hasHeader && header === undefined,
                                    'rux-dialog__header': true,
                                }}
                                part="header"
                            >
                                <slot
                                    name="header"
                                    onSlotchange={this._handleSlotChange}
                                >
                                    {header}
                                </slot>
                            </header>

                            <div class="rux-dialog__content">
                                <div
                                    class={{
                                        hidden:
                                            !this.hasMessage &&
                                            message === undefined,
                                        'rux-dialog__message': true,
                                    }}
                                    part="message"
                                >
                                    <slot onSlotchange={this._handleSlotChange}>
                                        {message}
                                    </slot>
                                </div>
                            </div>
                            <footer
                                class={{
                                    'rux-dialog__footer': true,
                                }}
                                part="footer"
                            >
                                {this.hasFooter ? (
                                    <slot
                                        name="footer"
                                        onSlotchange={this._handleSlotChange}
                                    ></slot>
                                ) : (
                                    <Fragment>
                                        <rux-button-group h-align="right">
                                            <rux-button
                                                secondary={
                                                    confirmText.length > 0
                                                }
                                                onClick={_handleDialogChoice}
                                                data-value="false"
                                                hidden={!denyText}
                                                tabindex="0"
                                                exportparts="container:deny-button"
                                                id="rux-dialog-deny-button"
                                            >
                                                {denyText}
                                            </rux-button>
                                            <rux-button
                                                onClick={_handleDialogChoice}
                                                hidden={!confirmText}
                                                data-value="true"
                                                tabindex="1"
                                                exportparts="container:confirm-button"
                                            >
                                                {confirmText}
                                            </rux-button>
                                        </rux-button-group>
                                        <slot
                                            name="footer"
                                            onSlotchange={
                                                this._handleSlotChange
                                            }
                                        ></slot>
                                    </Fragment>
                                )}
                            </footer>
                        </dialog>
                    </div>
                </Host>
            )
        )
    }
}
