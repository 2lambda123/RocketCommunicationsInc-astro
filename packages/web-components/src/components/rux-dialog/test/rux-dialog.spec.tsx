import { newSpecPage } from '@stencil/core/testing'
import { RuxDialog } from '../rux-dialog'

describe('rux-dialog', () => {
    it('builds', async () => {
        const modal: RuxDialog = new RuxDialog()
        const { open, modalMessage, modalTitle, confirmText, denyText } = modal
        expect(modal).toBeTruthy()
        expect({
            open,
            modalMessage,
            modalTitle,
            confirmText,
            denyText,
        }).toEqual({
            open: false,
            modalMessage: undefined,
            modalTitle: undefined,
            confirmText: 'Confirm',
            denyText: 'Cancel',
        })
    })
    it('renders with slots', async () => {
        const page = await newSpecPage({
            components: [RuxDialog],
            html: `
            <rux-dialog>
                <span slot="header">Header</span>
                <div slot="message">Message</div>
                <div slot="footer">Footer</div>
            </rux-dialog>
            `,
        })
        expect(page.root).toMatchSnapshot()
    })
    it('renders with a mix of slots and props', async () => {
        const page = await newSpecPage({
            components: [RuxDialog],
            html: `
            <rux-dialog modal-title="Title">
                <div slot="message">Message</div>
                <div slot="footer">Footer</div>
            </rux-dialog>
            `,
        })
        expect(page.root).toMatchSnapshot()
    })
})
