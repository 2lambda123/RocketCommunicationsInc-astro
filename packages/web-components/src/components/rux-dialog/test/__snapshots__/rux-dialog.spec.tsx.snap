// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-dialog renders with a mix of slots and props 1`] = `
<rux-dialog modal-title="Title">
  <mock:shadow-root></mock:shadow-root>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-dialog>
`;

exports[`rux-dialog renders with slots 1`] = `
<rux-dialog>
  <mock:shadow-root></mock:shadow-root>
  <span slot="header">
    Header
  </span>
  <div slot="message">
    Message
  </div>
  <div slot="footer">
    Footer
  </div>
</rux-dialog>
`;
